### ISSUE TITLE
[Push Notification] Extra data on payload

### ISSUE THREAD

**Issue by akamuraasai** [ConnectyCube/connectycube-reactnative-samples#27]
Hi, I'm using the Offline Notifications feature, and it's working fine, but I need to send some extra info into the notification payload and don't know how.
I'll explain the motivation behind this.

### Scenario
Like in any major chat app, when a user press on notification, it will navigate the app to the specific dialog where the notification came from.
There are 3 possibilities here:
- The first is that app is running in background and the dialog is already into the "redux store". So in this scenario, the only thing that I need to do is filter the dialogs list based on dialog_id that came into the notification and navigate to it.
- The second is that app is running in background and the dialog didn't  is into the "redux store", in this case, I just need to update the dialogs list and navigate to it, a task that runs pretty fast and isn't a problem at all.
- In the last scenario this isn't true. In this case the app is killed and the dialog is new, and didn't is into the "redux store", this is a very common scenario in my app and the worst. Here is why, the app is killed, so first I need to execute all the login+connect functions, wait until all this is complete and after all this stuff, update my dialogs list and based on the new list navigate to the correct dialog. This simple takes too long.

### Solution
So to fix this issue and make this navigation as fluid as possible, all I need is a way to send more info about the dialog, as title, description, type, XMPP JID and dialog User ID.
With all this info I can make all the listeners and other things work even if the app isn't fully loaded.

### Suggestion
It will be really good if I can send some extra data with the message send function with some special property like the `attachment`, or any other property like that. This will make this feature work even without break any other app that already uses the ConnectyCube SDK.

Any chance to make this work? Or it's a feature that already exists and I'm missing something?

**Comment by DaveLomber** [comment]
Hi @akamuraasai 

let me try to suggest a little bit diff solution, which can be done w/o an payload modifications 

As we already have a 'dialog_id' in a push payload, we can do/assume the following: 

First of all, a new dialog creation is not a frequent operation in standard chat apps. Usually, users are mostly send/receive messages, but not create dialogs.  

Anyway, if your app is killed and you need to get a dialog, you can do one the following:

-  on push received, you can run a background js task, to retrieve a dialog from server. There you can do only 1 or 2 requests - create session if it's expired already and retrieve a dialog. It's pretty fast. You do not need to connect to real time chat here.
- OR, on push click, after an app is opened, you can do the same  - make a request to server and retrieve a dialog, then store to redux and re-renader the UI (as this API request will take 1-2 sec, so there should not be any big deal for user). Also, you can create a dummy dialog in redux, with just an ID and navigate immediately a user there, and then, once a request from server is received, store a complete dialog in redux and re-render, so only dialog's data like name will be re-rendered, but a user will already have an ability to send-rececve messages  

Please let me know if anything from it works for you

**Comment by akamuraasai** [comment]
Hi, @DaveLomber.
About your suggestions I'm already using the second one. I just navigate the user to a blank chat screen with only the `dialog_id` and lazy load all the dialog data. This really takes only about 1 to 2 seconds in a WiFi or good LTE connection. But unfortunately this is not the connection that 80% of my users have. I'm making a app that operates on a region that you can expect the user to have a really bad EDGE connection, and even no connection at some time.
So even if this solution works and is ok, for most of my users this implies a 10 seconds loading screen or so, and in some cases an infinity loading, since it can lose the connection before the request ends.

Summing up, the second solution works, but it can fail in ways that hurts my users experiences. A way to send all the necessary dialog data along with the message is the best way to make the scenario works, even if the user don't have any internet at all, since my app supports offline messages.

**Comment by DaveLomber** [comment]
@akamuraasai  could you please describe in more details your use case, I may find a solution:

- why you need to create many dialogs, frequently 
- on open by push - do you need to navigate directly to chat screen, or to dialog profile?

**Comment by akamuraasai** [comment]
Sure, I can describe my use case, but I don't know if it's really necessary. Well, I work on a marketplace app that happens to have a chat feature. The chat is used to make a bridge between the customers and the sellers.
When a customer needs more information about a specific item, or some information about delivery times and that kind of stuff, he simply clicks on "Chat" button and a new Dialog including the customer, the seller and a bot is created so they can communicate whatever way they want.
A new dialog is a really normal thing here, since for every new seller the customer tries to make contact, a new dialog needs to be created. I can expect that a median single customer can have up to 30 dialogs. For the seller, every day almost 20 customers buys from them, and the list of dialogs can be tremendous. For each of this dialogs, I have the problem that I explained before.
Even worst is that the seller itself works in a place here in São Paulo that has a REALLY poor internet connection. So yes, the lazy loading solution works, but is far from ideal.
In apps like WhatsApp, Telegram and Facebook Messenger, when a new dialog is created and you navigate to it using a press on some push notifications, you'll immediately see the Dialog infos and can right away start to send messages. Of course it's not a fair comparison, but is something like that what I'm trying to achieve.

About your last question, I need the user to navigate directly to the chat screen, where he can start sending messages. The reason that the fields I've said before are:
- Title - To show on chat screen header.
- Type - To know if I need to use the chat title or user name on chat screen header, and to use on sendMessage function.
- XMPP JID - To use on sendMessage function.
- Dialog User ID - To use in typing listener, sendMessage function and etc.

I've think that it's all summarized now, let me know if you need some more information or have another idea to how to solve this problem.
Thanks for all the replies so far.

**Comment by DaveLomber** [comment]
@akamuraasai 

ok thanks for the provided information ,
indeed it's all a valid case 

I have a solution for you

Along with dialog creation, you can initiate a custom push via 'createEvent' API
https://developers.connectycube.com/js/push-notifications?id=send-push-notifications

and here you can include any amount of custom params in payload, 

This will work better I think. 

The way you proposed is also a valid one, but what I do not like is to include this all data on every chat message push to offline user, as in reallity this info is needed one in 1st chat message, 
so custom push event should be better here and you have more control 

PLease let me know if it will work for you

**Comment by akamuraasai** [comment]
Unfortunately, this is not a valid suggestion, @DaveLomber. It would be a perfect solution, but the data on push notifications are only available in the app if the user press it.
So if I send a push, telling the user that he has a new dialog, and then another push comes from this new dialog and the user press the later, all the the dialog info will still be missing, since the user pressed the “wrong” push notification. And ever worst is that push notification have not a 100% delivery warranty. So if the users phone happens to be in low battery or something, the custom push notification can never arrive.

Because of all this little annoying details that I’m asking to send more data on every push, I’ve worked with this kind of stuff several times and this is the only way to solve the problem AFAIK.

Any other ideas?

**Comment by akamuraasai** [comment]
Hi @DaveLomber, any news about this issue?

**Comment by ccvlad** [comment]
Since there is no activity I’m going to close the issue. Please raise a separate issue in case of any other questions.

**Comment by CMLCNL** [comment]
@akamuraasai did you find any solution? connectycube documentation trash..

**Comment by DaveLomber** [comment]
@CMLCNL what extra payload are you looking to add? Please describe your use case

**Comment by DaveLomber** [comment]
@CMLCNL the initiating a call - you are totally free to create a push event by yourself and add any data https://developers.connectycube.com/js/videocalling?id=receive-a-call-in-background

**Comment by CMLCNL** [comment]
i solved my issue..

  ```
NSObject* aps = payload.dictionaryPayload[@"aps"];
  
  NSString* alert = [aps valueForKey:@"alert"];
  
  NSLog(@"%@", alert);
  
  NSData *webData = [alert dataUsingEncoding:NSUTF8StringEncoding];
  NSError *error;
  NSDictionary *jsonDic = [NSJSONSerialization JSONObjectWithData:webData options:NSJSONReadingMutableContainers error:&error];
  
  
  NSLog(@"%@", jsonDic);
  
  
  NSString *uuid = [jsonDic valueForKey:@"uuid"];
  NSString *callerName = [jsonDic valueForKey:@"callerName"];
  NSString *handle = [jsonDic valueForKey:@"handle"];
```

**Comment by DaveLomber** [comment]
@CMLCNL  great it finally works at your side!
